File: PythonToJava/sort.py

# swap two elements of an array
def swap(a, idx1, idx2):
    t = a[idx1]
    a[idx1] = a[idx2]
    a[idx2] = t


# return the index of the minimum element in
# the subrange [lo,hi) of list a.
def min_element(a, lo, hi):
    min_idx = lo
    for i in range(lo+1, hi):
        if a[i] < a[min_idx]:
            min_idx = i
    return min_idx


# sort the given list using the selection sort algorithm
def selection_sort(a):
    for i in range(0,len(a)):
        idx = min_element(a, i, len(a))
        swap(a, i, idx)


# sort the given list using the insertion sort algorithm
def insertion_sort(a):
    for i in range(1,len(a)):
        j = i
        while (j > 0 and a[j] < a[j-1]):
            swap(a, j, j-1)
            j -= 1


# test the selection_sort and insertion_sort functions
if __name__ == "__main__":
    a1 = [ 27, 82, 41, 124, 62, 31, 94 ]

    a = a1[:]
    selection_sort(a)
    print(a)

    a = a1[:]
    insertion_sort(a)
    print(a)

    weasleys = [ "Bill", "Charlie", "Percy", "George", "Fred", "Ron", "Ginny" ]
    insertion_sort(weasleys)
    print(weasleys)

File: PythonToJava/sort.java
public class sort {
    
    // swap two elements of the given array
    public static void swap(int[] a, int idx1, int idx2) {
        int t = a[idx1];
        a[idx1] = a[idx2];
        a[idx2] = t;
    }

    // return the index of the minimum element in the
    // [lo,hi) subrange of array a.
    public static int minElement(int[] a, int lo, int hi) {
        int minIdx = lo;
        for (int i = lo+1; i < hi; i++)
            if (a[i] < a[minIdx])
                minIdx = i;
        return minIdx;
    }

    // sort the given array using the selection sort algorithm
    public static void selectionSort(int[] a) {
        for (int i = 0; i < a.length; i++) {
            int idx = minElement(a, i, a.length);
            swap(a, i, idx);
        }
    }

    // sort the given list using the insertion sort algorithm
    public static void insertionSort(int[] a) {
        for (int i = 1; i < a.length; i++) {
            int j = i;
            while (j > 0 && a[j] < a[j-1]) {
                swap(a, j, j-1);
                j -= 1;
            }
        }
    }

    // create a copy of an array
    public static int[] copyArray(int[] a) {
        int[] b = new int[a.length];
        for (int i = 0; i < a.length; i++)
            b[i] = a[i];
        return b;
    }

    // print the contents of an array
    public static void printArray(int[] a) {
        System.out.printf("[ %d", a[0]);
        for (int i = 1; i < a.length; i++)
            System.out.printf(", %d", a[i]);
        System.out.println(" ]");
    }

    // test the selection_sort and insertion_sort functions
    public static void main(String[] args) {
        int[] a = { 27, 82, 41, 124, 62, 31, 94 };

        int[] b = copyArray(a);
        selectionSort(b);
        printArray(b);

        b = copyArray(a);
        insertionSort(b);
        printArray(b);

        // String[] weasleys = { "Bill", "Charlie", "Percy", "George", "Fred", "Ron", "Ginny" };
        // insertionSort(weasleys);
        // printArray(weasleys);
    }
}
